### PR TITLE
go: do not replace path to zoneinfo.zip and mime.types

### LINKED ISSUE #54603: Go: update path to zoneinfo.zip
## Issue description

A Go program compiled using nix-shell returns an error when running on a separate server and trying to load a timezone file.

### Steps to reproduce

1. write `main,go` file with:

```
package main

import (
	"log"
)

func main() {
	l, err := time.LoadLocation("America/New_York")
	if err != nil {
		log.Fatalln("failed to read timezone file: ", err)
	}
	log.Println("All went well: " + l.String())
}
```

default.nix:

```
# This pins the packages to certain versions
with import (builtins.fetchTarball {
  # Descriptive name to make the store path easier to identify
  name = "nixos-unstable-2019-01-22";
  # Commit hash for nixos-unstable as of 2019-01-22
  url = https://github.com/NixOS/nixpkgs/archive/51d88f7d20ea6f992934927539bd3caea28bcec6.tar.gz;
  # Hash obtained using `nix-prefetch-url --unpack <url>`
  sha256 = "198c54x34c03pv48xq7sh3b81q5wl0h3g7gq240s31q56zc6jpr3";
}) {};

stdenv.mkDerivation rec {
  name = "env";
  env = buildEnv { name = name; paths = buildInputs; };
  buildInputs = [
    go
  ];
}
```

$nix-shell
$<inside shell>go build
$exit
$ ./app # name of the app you just created
2019/01/25 20:04:58 All went well: America/New_York

Now copy this file to another computer

#./app
2019/01/26 01:01:34 could not load new york locationopen /nix/store/0596k27r6pacyx4jp1bcn8zqyjhji52l-go-1.11.4/share/go/lib/time/zoneinfo.zip: no such file or directory

The error includes the zip name due to internal reasons the Go team has, but the real problem is that nix replaced 

`/usr/share/zoneinfo/`
with
`/nix/store/bgzqg5q8q1gpsbmpy4mavvnjzy528c8l-tzdata-2018g/share/zoneinfo/`

I would be more than happy to send a Pull Request if you agree with the proposed solution of prepending instead of replacing.

## Technical details

$ nix-shell -p nix-info --run "nix-info -m"
these paths will be fetched (0.00 MiB download, 0.00 MiB unpacked):
  /nix/store/dzvh4f6ihkwg44mmf2awxg564k1s8khg-nix-info
copying path '/nix/store/dzvh4f6ihkwg44mmf2awxg564k1s8khg-nix-info' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 4.4.0-141-generic, Ubuntu, 16.04.5 LTS (Xenial Xerus)`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0`
 - channels(diego): `"nixpkgs-18.03pre130569.7a04c2ca296"`
 - nixpkgs: `/home/diego/.nix-defexpr/channels/nixpkgs`

version is nixos-unstable-2019-01-22

The file `pkgs/development/compilers/go/1.11.nix` replaces the path that the go runtime looks for the timezone information.

Original Go code is:

```
var zoneSources = []string{
        "/usr/share/zoneinfo/",
        "/usr/share/lib/zoneinfo/",
        "/usr/lib/locale/TZ/",
        runtime.GOROOT() + "/lib/time/zoneinfo.zip",
```

But nix changes the first entry, `"/usr/share/zoneinfo/"` for `${tzdata}/share/zoneinfo/`

This works fine if you end up running your go application in the same nix env you compiled it, but when you deploy your go program to a separate server, the path, that can look like `/nix/store/bgzqg5q8q1gpsbmpy4mavvnjzy528c8l-tzdata-2018g/share/zoneinfo/`

isn't present on the separate server and then the lookup fails.

`"/usr/share/zoneinfo/"` is a very common path in Linux distributions. A better way to solve this would be to **prepend** the nix only path, as the first entry, and then keep `"/usr/share/zoneinfo/"` and the other entries as they are in the original go source code.

The benefits are that those who compile and run in the same nix environment will use the nix path, but those who deploy somewhere else won't have to do any extra work.

### PR DESCRIPTION
**Pull request by fmpwizard**
Prepend the nix path to the zoneinfo.zip file and keep the original alternatives
to allow go programs built using nix to run on non nix servers.

see https://github.com/NixOS/nixpkgs/issues/54603

###### Motivation for this change

Fixes: #54603 

Allow Go programs built using nix to run on servers that don't have nix

###### Things done

I am very new to nix and could not figure out how to test this branch locally, reading 
https://nixos.org/nixpkgs/manual/#chap-submitting-changes

I don't know what to replace this like with:

```
nix-env -i pkg-name -f <path to your local nixpkgs folder>
```

```
$ nix-env -i go -f /nix/store

[...] many warning similar to:
warning: name collision in input Nix expressions, skipping '/nix/store/mcgh4v55zwfp6f1x4b4zrdp6lrzki475-user-environment/nixpkgs'
[...] and finally:
error: cannot auto-call a function that has an argument without a default value ('derivations')

```

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @veprbl

### CODE REVIEW

**Comment by Mic92** [comment]
The `-f` parameter should point to the git repository you used to make the pull request from: 

```console
$ git clone https://github.com/fmpwizard/nixpkgs
$ cd nixpkgs
$nixpkgs> git checkout issue_54603
$nixpkgs> nix-env -i go -f .
```

**Comment by Mic92** [comment]
@GrahamcOfBorg build go

**Comment by kalbasit** [comment]
The same thing is happening with the mime types. I suggest fixing that as well if we want full support outside of Nix.

Our side: https://github.com/NixOS/nixpkgs/blob/5a0be78562c7de3daa450ee2c4698842e8ff7f8c/pkgs/development/compilers/go/1.13.nix#L59-L61

Upstream code: https://github.com/golang/go/blob/da4d58587e0e4028ea384580053c3c455127e446/src/mime/type_unix.go#L19-L23

**Comment by fmpwizard** [comment]
@kalbasit great idea, I added a new commit. I tried to test this locally by running

    nix-env -i go -f .

(at the root of this repo)

but I get this error:

```
--- FAIL: TestChown (0.00s)
        os_unix_test.go:51: gid: 100
        os_unix_test.go:63: groups:  [65534 100]
        os_unix_test.go:66: chown /tmp/_Go_TestChown141509686 -1 65534: chown /tmp/_Go_TestChown141509686: invalid argument
FAIL

```

I'm on Fedora 30 5.3.14-200.fc30.x86_64 

Thanks!

**Comment by kalbasit** [comment]
it works locally for me. Can you please fix the commit, it should `go: do not replace path to mime.types`. It must start with `go: ` (see [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md)) and the rest is for consistency. Please also update the title of the PR to reflect the change.

Once that's done, I'll submit it for build by ofborg and test locally on Linux/Mac.

**Comment by fmpwizard** [comment]
@kalbasit thanks, I updated the commit message and the PR title

**Comment by kalbasit** [comment]
@GrahamcOfBorg build go pet jx

File: pkgs/development/compilers/go/1.13.nix [hunk 1]
```diff
@@ -97,7 +97,10 @@ stdenv.mkDerivation rec {
     rm src/net/cgo_unix_test.go
 
   '' + optionalString stdenv.isLinux ''
-    sed -i 's,/usr/share/zoneinfo/,${tzdata}/share/zoneinfo/,' src/time/zoneinfo_unix.go
+    # prepend the nix path to the zoneinfo files but also leave the original value for static binaries
+    # that run outside a nix server
+    sed -i 's,\"/usr/share/zoneinfo/,"${tzdata}/share/zoneinfo/\"\,\n\t&,' src/time/zoneinfo_unix.go
```
Comment (kalbasit) [inline pkgs/development/compilers/go/1.13.nix]: I wonder why are using the zoneinfo from Nix on Linux only. Is there something I'm not aware of @LnL7?

**Review by kalbasit — APPROVED** [review]

**Pull request merged**